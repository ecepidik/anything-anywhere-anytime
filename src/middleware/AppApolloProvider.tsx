import React, { FunctionComponent, PropsWithChildren } from 'react';

import { ApolloClient, InMemoryCache, ApolloProvider, from, ApolloLink } from '@apollo/client';
import { onError } from '@apollo/client/link/error';
import { createNetworkStatusNotifier } from 'react-apollo-network-status';
import createUploadLink from 'apollo-upload-client/createUploadLink.mjs';

const unauthorizedInterceptor = onError(({ networkError }) => {
    console.log('Unauthorized');
})
const { useApolloNetworkStatus, link: networkStatusNotifierLink } = createNetworkStatusNotifier();
// eslint-disable-next-line no-restricted-globals
const uploadLink: ApolloLink = networkStatusNotifierLink.concat(createUploadLink({uri: `${location.protocol}//${location.host}/graphql`}));
// const uploadLink = createUploadLink({uri: `${location.protocol}//${location.host}/graphql`});

const link = from([uploadLink, unauthorizedInterceptor])

const apolloClient = new ApolloClient({
    link: link,
    cache: new InMemoryCache(),
  });

  export const globalPendingApolloMutations = () => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const status = useApolloNetworkStatus();
    return status.numPendingMutations;
  }

  export const AppApolloProvider: FunctionComponent<PropsWithChildren<unknown>> = ({ children }) => {
    return <ApolloProvider client={apolloClient}>{children}</ApolloProvider>
  }
