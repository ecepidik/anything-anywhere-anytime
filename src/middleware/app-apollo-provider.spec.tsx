/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/render-result-naming-convention */
import React from 'react';
import { render } from '@testing-library/react';
import { AppApolloProvider } from './AppApolloProvider';

jest.mock('@apollo/client', () => ({
    ApolloClient: class {
        constructor(public mockApolloConfig: Record<string, unknown>) {}
    },
    ApolloLink: class {
        constructor(public mockLinkConfig: Record<string, unknown>) {}
    },
    ApolloProvider: 'mock-apollo-provider',
    InMemoryCache: class {
        constructor(public mockCacheConfig: Record<string, unknown>) {}
    },
    from: (middlewares: []) => middlewares,
    gql: () => 'graphql'
}));

jest.mock('react-apollo-network-status', () => ({
    createNetworkStatusNotifier: jest.fn(() => ({
        link: {
            concat: jest.fn(_props => 'test-link')
        }
    }))
}));

// jest.mock('apollo-upload-client/createUploadLink.mjs', () => ({
//     createUploadLink: () => 'link'
// }));

jest.mock('apollo-upload-client/createUploadLink.mjs', () => ({
    __esModule: true,
  default: jest.fn(),
}))

describe('testing app apollo provider', () => {
    it('should create apollo client with configuration', () => {
        const component = render(
            <AppApolloProvider>
                <div>
                    children
                </div>
            </AppApolloProvider>
        );
        expect(component.container.firstChild).toMatchSnapshot();
    })
})