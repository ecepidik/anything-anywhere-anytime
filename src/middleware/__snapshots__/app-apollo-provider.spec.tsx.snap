// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing app apollo provider should create apollo client with configuration 1`] = `
<mock-apollo-provider
  client="[object Object]"
>
  <div>
    children
  </div>
</mock-apollo-provider>
`;
